### PR TITLE
Error tree / branched errors

### PR DESCRIPTION
**Pull request by kurtbrose**
this should fix that weird glitch where the last error was "sticky"

still need to add formatting code for the side-branches

### CODE REVIEW

File: glom/core.py [hunk 1]
```diff
@@ -1832,11 +1845,13 @@ def glom(target, spec, **kwargs):
...
-    parent[LAST_CHILD_SCOPE] = scope
-    scope[T] = target
-    scope[Spec] = spec
-    scope[UP] = parent
+    scope = scope.new_child({
+        T: target,
+        Spec: spec,
+        UP: parent,
+        CHILD_ERRORS: [],
+    })
```
Comment (mahmoud) [inline glom/core.py:1853]: This reads better, nice.

File: glom/core.py [hunk 2]
```diff
@@ -176,7 +176,19 @@ def _unpack_stack(scope):
     #
     # root glom call generates two scopes up at the top that aren't part
     # of execution
-    return [(scope, scope[Spec], scope[T]) for scope in list(reversed(scope.maps[:-2]))]
+    stack = []
```
Comment (mahmoud) [inline glom/core.py:179]: can't comment there, but you need to update the docstring

Comment (mahmoud) [reply glom/core.py:179]: also probably note in there the order of the output frames

File: glom/core.py [hunk 3]
```diff
@@ -1821,7 +1834,7 @@ def glom(target, spec, **kwargs):
         else:
             err = GlomError.wrap(e)
         if isinstance(err, GlomError):
-            err._finalize(scope[ERROR_SCOPE])
+            err._finalize(scope[LAST_CHILD_SCOPE])
```
Comment (mahmoud) [inline glom/core.py:1837]: Is `LAST_CHILD_SCOPE` a misnomer? As in, it's not the scope of the last child in the spec, it's actually the most-recently-evaluated spec? Not sure this helps much but would `LATEST_CHILD_SCOPE` be more accurate? More runtimey?

File: glom/core.py [hunk 4]
```diff
@@ -216,7 +228,7 @@ def format_oneline_trace(scope):
     # if the target is the same, don't repeat it
     segments = []
     prev_target = _MISSING
-    for scope, spec, target in _unpack_stack(scope):
+    for scope, spec, target, branches in _unpack_stack(scope):
```
Comment (mahmoud) [inline glom/core.py:231]: so based on the PR description, there's at most one branch here?

File: glom/core.py (hunk 3 above)
Comment (kurtbrose) [reply glom/core.py:1837]: yes, I'm trying to say "most recent" child

File: glom/core.py (hunk 4 above)
Comment (kurtbrose) [reply glom/core.py:231]: it will need to do a recursive call with depth=depth+1 on each  branch

File: glom/core.py (hunk 1 above)
Comment (kurtbrose) [reply glom/core.py:1853]: nice, I mostly did it actually to unpack the function a little -- I felt bad about adding even more stuff to _glom() and played with timeit, discovered that the __setitem__ calls on the ChainMap completely dominate the list construction time

building the dict constant like this and passing it in is much, much faster than the other way -- like 5-10x as fast

Comment (mahmoud) [reply glom/core.py:1853]: totally makes sense, I was thinking the same, but didn't wanna prematurely optimize :)

File: glom/core.py (hunk 3 above)
Comment (mahmoud) [reply glom/core.py:1837]: Would "latest", "newest", or "current" be clearer?

File: glom/core.py (hunk 4 above)
Comment (mahmoud) [reply glom/core.py:231]: hmm, I guess I might be asking a simpler question: `branches` can be an empty list or a list with one item?

**Comment by kurtbrose** [comment]
here's v1

```
Traceback (most recent call last):
  File "test_error.py", line ___, in _make_stack
    glom(target, spec)
  File "core.py", line ___, in glom
    raise err
glom.matching.MatchError: error raised while processing, details below.
 Target-spec trace (most recent last):
 - Target: [None]
 - Spec: Match(Switch([(1, 1), ('a', 'a'), (T.a, T.a)]))
  Failed Branch:
   - Spec: 1
   [None] does not match 1
  Failed Branch:
   - Spec: 'a'
   [None] does not match 'a'
 - Spec: Switch([(1, 1), ('a', 'a'), (T.a, T.a)])
 - Spec: T.a
glom.matching.MatchError: no matches for target in Switch
```

**Comment by kurtbrose** [comment]
I need to figure out how to integrate the error messages... v2 (don't think I like how this came out):

```
Traceback (most recent call last):
  File "test_error.py", line ___, in _make_stack
    glom(target, spec)
  File "core.py", line ___, in glom
    raise err
glom.matching.MatchError: error raised while processing, details below.
 Target-spec trace (most recent last):
 - Target: [None]
 - Spec: Match(Switch([(1, 1), ('a', 'a'), (T.a, T.a)]))
   Failed Branch ([None] does not match 1):
   - Spec: 1
   Failed Branch ([None] does not match 'a'):
   - Spec: 'a'
 - Spec: Switch([(1, 1), ('a', 'a'), (T.a, T.a)])
 - Spec: T.a
glom.matching.MatchError: no matches for target in Switch
```

**Comment by kurtbrose** [comment]
I think to try and make it look like a normal stack trace --

just hand generate what I want it to look like

```
Traceback (most recent call last):
  File "test_error.py", line ___, in _make_stack
    glom(target, spec)
  File "core.py", line ___, in glom
    raise err
glom.matching.MatchError: error raised while processing, details below.
 Target-spec trace (most recent last):
 - Target: [None]
 - Spec: Match(Switch([(1, 1), ('a', 'a'), (T.a, T.a)]))
   Failed Branch:
   - Spec: 1
   glom.matching.MatchError: [None] does not match 1
   Failed Branch:
   - Spec: 'a'
   glom.matching.MatchError: [None] does not match 'a'
 - Spec: Switch([(1, 1), ('a', 'a'), (T.a, T.a)])
 - Spec: T.a
 glom.core.PathAccessError: could not access 'a', part 0 of T.a, got error: AttributeError("'list' object has no attribute 'a'")
glom.matching.MatchError: no matches for target in Switch
```

File: glom/core.py (hunk 3 above)
Comment (kurtbrose) [reply glom/core.py:1856]: I dunno, its been around for awhile

```
LAST_CHILD_SCOPE.__doc__ = """
Marker that can be used by parents to keep track of the last child
scope executed.  Useful for "lifting" results out of child scopes
for scopes that want to chain the scopes of their children together
similar to tuple.
"""
```

I guess this is a new use case but renaming seems relatively low priority

Comment (kurtbrose) [reply glom/core.py:1856]: I guess the difference is if you consider the child completed or still in progress

File: glom/core.py (hunk 4 above)
Comment (kurtbrose) [reply glom/core.py:250]: empty list

Comment (kurtbrose) [reply glom/core.py:250]:
```python
branches = scope.maps[0][CHILD_ERRORS][:-1]
```

File: glom/test/test_error.py [hunk 5]
```diff
@@ -230,6 +230,27 @@ def __repr__(self):
...
+    raise err
+glom.matching.MatchError: error raised while processing, details below.
+ Target-spec trace (most recent last):
+ - Target: [None]
+ - Spec: Match(Switch([(1, 1), ('a', 'a'), (T.a, T.a)]))
+ - Spec: Switch([(1, 1), ('a', 'a'), (T.a, T.a)])
+   Failed Branch ([None] does not match 1):
+   - Spec: 1
+   Failed Branch ([None] does not match 'a'):
+   - Spec: 'a'
```
Comment (mahmoud) [inline glom/test/test_error.py:249]: so I see `'a'` duplicated here, when would it not be? I'm thinking maybe we change the bullet notation (currently always `-`) for branches to be kind of "connected", more tersely showing that things are done in parallel maybe? Like, actual tree notation?

Comment (kurtbrose) [reply glom/test/test_error.py:249]: the first 'a' is coming from `str(MatchError)`, the second is coming from `repr(spec)`

Comment (kurtbrose) [reply glom/test/test_error.py:249]: are you thinking an ascii art tree?

**Comment by kurtbrose** [comment]
so this is kind of interesting:

```python
glom.core.CoalesceError: error raised while processing, details below.
 Target-spec trace (most recent last):
 - Target: {'a': {'b': 'c'}, 'd': {'e': ['f'], 'g': 'h'}, 'i': [{'j... (len=4)
 - Spec: Coalesce('xxx', 'yyy')
  Failed Branch:
   - Spec: 'xxx'
  PathAccessError: could not access 'xxx', part 0 of Path('xxx'), got error: KeyError('xxx')
 - Spec: 'yyy'
glom.core.CoalesceError: no valid values found. Tried ('xxx', 'yyy') and got (PathAccessError, PathAccessError) (at path [])
```

**Comment by kurtbrose** [comment]
some thoughts

1- we have tests that check glom error formats -- we should probably make them into these stack trace tests since it is way easier to audit in that form

2- case in point of 1: `(at path [])` -- I think that comes from here `                raise CoalesceError(self, skipped, scope[Path])`; so that mechanism is a little busted right now

3- its neat how `Coalesce` gets that structure around it automatically -- and maybe `Coalesce` no longer needs to work as hard with packing all the details into the error string

**Comment by kurtbrose** [comment]
its a bit messy, but look at this:

```python
Traceback (most recent call last):
  File "test_error.py", line ___, in _make_stack
    glom(target, spec)
  File "core.py", line ___, in glom
    raise err
glom.matching.MatchError: error raised while processing, details below.
 Target-spec trace (most recent last):
 - Target: [None]
 - Spec: Match(Switch([(1, 1), ('a', 'a'), (T.a, T.a)]))
 - Spec: Switch([(1, 1), ('a', 'a'), (T.a, T.a)])
glom.matching.MatchError: no matches for target in Switch
  Failed Branch:
   - Spec: 1
  glom.matching.MatchError: [None] does not match 1
  Failed Branch:
   - Spec: 'a'
  glom.matching.MatchError: [None] does not match 'a'
 - Spec: T.a
PathAccessError: could not access 'a', part 0 of T.a, got error: AttributeError("'list' object has no attribute 'a'")
glom.matching.MatchError: no matches for target in Switch
```

so it prints the MatchError AGAIN -- but, this may you kind of see the errors that happened down each path, and then at the level they all come together you see "no matches in switch"

(something similar happens for coalesce)

**Comment by kurtbrose** [comment]
looking at this, it almost seems that if it was a multi-branch, all the children should be treated as equal rather than the last one getting special treatment

**Comment by kurtbrose** [comment]
here's the latest, for a decently nested spec

```python
Match(Switch([
    (1, 1),
    ('a', 'a'),
    ([None], Switch([
       (1, 1),
       ('a', 'a'),
       ([None], T.a)]))]))
```



```python
glom.core.PathAccessError: error raised while processing, details below.
 Target-spec trace (most recent last):
 - Target: [None]
 - Spec: Match(Switch([(1, 1), ('a', 'a'), ([None], Switch([(1, 1), ('a', '...
 - Spec: Switch([(1, 1), ('a', 'a'), ([None], Switch([(1, 1), ('a', 'a'), (...
  Failed Branch:
   - Spec: 1
   - glom.matching.MatchError: [None] does not match 1
  Failed Branch:
   - Spec: 'a'
   - glom.matching.MatchError: [None] does not match 'a'
  Failed Branch:
   - Spec: [None]
   - Spec: Switch([(1, 1), ('a', 'a'), ([None], T.a)])
   - glom.core.PathAccessError: could not access 'a', part 0 of T.a, got error: AttributeError("'list' object has no attribute 'a'")
    Failed Branch:
     - Spec: 1
     - glom.matching.MatchError: [None] does not match 1
    Failed Branch:
     - Spec: 'a'
     - glom.matching.MatchError: [None] does not match 'a'
    Failed Branch:
     - Spec: [None]
     - Spec: T.a
     - glom.core.PathAccessError: could not access 'a', part 0 of T.a, got error: AttributeError("'list' object has no attribute 'a'")
glom.core.PathAccessError: could not access 'a', part 0 of T.a, got error: AttributeError("'list' object has no attribute 'a'")
```

**Comment by kurtbrose** [comment]
next thing, I'll come up with a heuristic to avoid the repetitious printing of the last error

(maybe just pass in the root error and skip printing it?)

also, thinking about tree structure drawing

```python
 + Spec: Switch([(1, 1), ('a', 'a'), ([None], Switch([(1, 1), ('a', 'a'), (...
  |Failed Branch:
  |- Spec: 1
  |- glom.matching.MatchError: [None] does not match 1
  |Failed Branch:
  |- Spec: 'a'
  |- glom.matching.MatchError: [None] does not match 'a'
  |Failed Branch:
  |- Spec: [None]
  |+ Spec: Switch([(1, 1), ('a', 'a'), ([None], T.a)])
  |- glom.core.PathAccessError: could not access 'a', part 0 of T.a, got error: AttributeError("'list' object has no attribute 'a'")
   ||Failed Branch:
    ||- Spec: 1
    ||- glom.matching.MatchError: [None] does not match 1
    ||Failed Branch:
```

**Comment by kurtbrose** [comment]
alright no more double prints of the root error:

```python
Traceback (most recent call last):
  File "test_error.py", line ___, in _make_stack
    glom(target, spec)
  File "core.py", line ___, in glom
    raise err
glom.core.PathAccessError: error raised while processing, details below.
 Target-spec trace (most recent last):
 - Target: [None]
 - Spec: Match(Switch([(1, 1), ('a', 'a'), ([None], Switch([(1, 1), ('a', '...
 - Spec: Switch([(1, 1), ('a', 'a'), ([None], Switch([(1, 1), ('a', 'a'), (...
  Failed Branch:
   - Spec: 1
   - glom.matching.MatchError: [None] does not match 1
  Failed Branch:
   - Spec: 'a'
   - glom.matching.MatchError: [None] does not match 'a'
  Failed Branch:
   - Spec: [None]
   - Spec: Switch([(1, 1), ('a', 'a'), ([None], T.a)])
    Failed Branch:
     - Spec: 1
     - glom.matching.MatchError: [None] does not match 1
    Failed Branch:
     - Spec: 'a'
     - glom.matching.MatchError: [None] does not match 'a'
    Failed Branch:
     - Spec: [None]
     - Spec: T.a
glom.core.PathAccessError: could not access 'a', part 0 of T.a, got error: AttributeError("'list' object has no attribute 'a'")
```

**Comment by kurtbrose** [comment]
made dict keys the parent of dict vals (was thinking about that for a long time to make variable capture nicer)

just did it for match for now, but maybe should do the same for Auto, Fill, etc

as part of this, made `Required()` into an object specially handled by the match dict-handler -- if it was a standalone object, then it would count as the parent and make variable capture impossible

**Comment by kurtbrose** [comment]
2.7 is choking on unicode in some weird ways

```python
E       assert 'Traceback (m... \'value\'")\n' == 'Traceback (mo... \'value\'")\n'
E         Skipping 305 identical leading characters in diff, use -v to show
E         - ': [{'valué': 'value'}]}
E         ?           ^
E         + ': [{'valu\xe9': 'value'}]}
E         ?           ^^^^
E         -  - Spec: [{'valué': 'value'}]
E         ?                 ^...
E
E         ...Full output truncated (10 lines hidden), use '-vv' to show


E       AssertionError: assert 'Traceback (m...et in Switch\n' == 'Traceback (mo...t in Switch\n'
E         Skipping 308 identical leading characters in diff, use -v to show
E         - [(1, 1), ('a', 'a'), (T.a, T.a)]))
E         + [(1, 1), (\'a\', \'a\'), (T.a, T.a)]))
E         ?           +  +   +  +
E         -  - Spec: Switch([(1, 1), ('a', 'a'), (T.a, T.a)])
E         +  - Spec: Switch([(1, 1), (\'a\', \'a\'), (T.a, T.a)])
E         ?                           +  +   +  +...
E
E         ...Full output truncated (17 lines hidden), use '-vv' to show


E       assert 'Traceback (m...bute \'a\'")\n' == 'Traceback (mo...ute \'a\'")\n'
E         Skipping 309 identical leading characters in diff, use -v to show
E         - [(1, 1), ('a', 'a'), ([None], T.a)]))
E         + [(1, 1), (\'a\', \'a\'), ([None], T.a)]))
E         ?           +  +   +  +
E         -  - Spec: Switch([(1, 1), ('a', 'a'), ([None], T.a)])
E         +  - Spec: Switch([(1, 1), (\'a\', \'a\'), ([None], T.a)])
E         ?                           +  +   +  +...
E
E         ...Full output truncated (17 lines hidden), use '-vv' to show


E       AssertionError: assert 'Traceback (m...(at path [])\n' == 'Traceback (mo...at path [])\n'
E         Skipping 269 identical leading characters in diff, use -v to show
E         -  Target: {'a': {'b': 'c'}, 'd': {'e': ['f'], 'g': 'h'}, 'i': [{'j... (len=4)
E         +  Target: {\'a\': {\'b\': \'c\'}, \'d\': {\'e\': [\'f\'], \'g\': \'h\'}, \'i\': [{\'j... (len=4)
E         ?           +  +    +  +   +  +    +  +    +  +    +  +    +  +   +  +    +  +     +
E         -  - Spec: Coalesce('xxx', 'yyy')
E         +  - Spec: Coalesce(\'xxx\', \'yyy\')
E         ?                   +    +   +    +...
E
E         ...Full output truncated (18 lines hidden), use '-vv' to show


E       assert 'é' in 'Traceback (most recent call last):\n  File "test_error.py", line ___, in _make_stack\n    glom(target, spec)\n  File "co...re.PathAccessError: could not access \'a\', part 0 of Path(\'a\', \'\xe9\', \'i\', \'o\'), got error: KeyError(\'a\')\n'
```

**Comment by mahmoud** [comment]
I think the structure drawing above is going in the direction I was imagining, ideally it would get clear enough that we wouldn't need a dedicated line for "Failed branch", it would be visually obvious (or folding into the spec line.

**Comment by kurtbrose** [comment]
oooohhhhhhhhhh heh, I see

```python
    if not isinstance(formatted_stack, unicode):
        # lil hack for py2
        # note that we only support the one unicode character
        formatted_stack = formatted_stack.decode('utf8') .replace(r'\xc3\xa9', u'é')
        formatted_stack = re.sub(r'\bu"', '"', formatted_stack)
        formatted_stack = re.sub(r"\bu'", "'", formatted_stack)
```

^ by changing the encoding I broke that little hack

**Comment by kurtbrose** [comment]
no, that doesn't make sense -- changing the encoding of the file shouldn't change to unicode literals

**Comment by kurtbrose** [comment]
yea, ensuring the py2 hack runs has no effect -- also it wouldn't covered \' replacement anyway

```python

    if True:  # not isinstance(formatted_stack, unicode):
        # lil hack for py2
        # note that we only support the one unicode character
        formatted_stack = formatted_stack.decode('utf8') .replace(r'\xc3\xa9', u'é')
        formatted_stack = re.sub(r'\bu"', '"', formatted_stack)
        formatted_stack = re.sub(r"\bu'", "'", formatted_stack)
```

**Comment by kurtbrose** [comment]
I think this might be some pytest nonsense somehow -- there aren't any `\` in the original string

**Comment by kurtbrose** [comment]
oh yea, you can kind of see

```python
E         -  - Target: {'a': {'b': 'c'}, 'd': {'e': ['f'], 'g': 'h'}, 'i': [{'j... (len=4)
E         +  - Target: {\'a\': {\'b\': \'c\'}, \'d\': {\'e\': [\'f\'], \'g\': \'h\'}, \'i\': [{\'j... (len=4)
```

the second string is way longer, because the `\` are added waaaay after the fact; neither the input nor the return of normalize stack has that

**Comment by kurtbrose** [comment]
ok, latest error is this:

```python
E           assert 'Traceback (m...ror object>\n' == 'Traceback (mo...ute \'a\'")\n'
E             Skipping 174 identical leading characters in diff, use -v to show
E             - essError: <unprintable PathAccessError object>
E             + essError: error raised while processing, details below.
E             +  Target-spec trace (most recent last):
E             +  - Target: [None]
E             +  - Spec: Match(Switch([(1, 1), ('a', 'a'), ([None], Switch([(1, 1), ('a', '...
E             +  - Spec: Switch([(1, 1), ('a', 'a'), ([None], Switch([(1, 1), ('a', 'a'), (......
E             
E             ...Full output truncated (20 lines hidden), use '-vv' to show
```

**Comment by kurtbrose** [comment]
so some kind of exceptions was thrown while generating the stack trace.... but I can't repro locally, even with the docker image for the exact python 3.6.7 as is being used by travis ci

**Comment by kurtbrose** [comment]
```
# tox -e py36
GLOB sdist-make: /glom/setup.py
py36 create: /glom/.tox/py36
py36 installdeps: -rrequirements.txt
py36 inst: /glom/.tox/.tmp/package/1/glom-20.5.1.dev0.zip
py36 installed: atomicwrites==1.4.0,attrs==19.1.0,boltons==20.2.0,coverage==5.2.1,face==20.1.1,filelock==3.0.12,glom @ file:///glom/.tox/.tmp/package/1/glom-20.5.1.dev0.zip,importlib-metadata==1.7.0,more-itertools==8.4.0,pluggy==0.13.1,py==1.9.0,pytest==4.1.1,PyYAML==3.13,six==1.15.0,toml==0.10.1,tox==3.7.0,virtualenv==15.1.0,zipp==3.1.0
py36 run-test-pre: PYTHONHASHSEED='1108509831'
py36 runtests: commands[0] | coverage run --parallel --rcfile /glom/.tox-coveragerc -m pytest --doctest-modules /glom/.tox/py36/lib/python3.6/site-packages/glom
=========================================================================================== test session starts ============================================================================================platform linux -- Python 3.6.7, pytest-4.1.1, py-1.9.0, pluggy-0.13.1
cachedir: .tox/py36/.pytest_cache
rootdir: /glom, inifile: pytest.ini
collected 226 items

py36/lib/python3.6/site-packages/glom/core.py .................s.....                                                                                                                                [ 10%] py36/lib/python3.6/site-packages/glom/grouping.py .....s                                                                                                                                             [ 12%] py36/lib/python3.6/site-packages/glom/matching.py .........                                                                                                                                          [ 16%] py36/lib/python3.6/site-packages/glom/mutation.py .....                                                                                                                                              [ 19%] py36/lib/python3.6/site-packages/glom/reduction.py ......                                                                                                                                            [ 21%] py36/lib/python3.6/site-packages/glom/streaming.py ..............                                                                                                                                    [ 27%] py36/lib/python3.6/site-packages/glom/tutorial.py .                                                                                                                                                  [ 28%] py36/lib/python3.6/site-packages/glom/test/test_basic.py .....................                                                                                                                       [ 37%] py36/lib/python3.6/site-packages/glom/test/test_check.py ...                                                                                                                                         [ 38%] py36/lib/python3.6/site-packages/glom/test/test_cli.py .........                                                                                                                                     [ 42%] py36/lib/python3.6/site-packages/glom/test/test_error.py ...................                                                                                                                         [ 51%] py36/lib/python3.6/site-packages/glom/test/test_fill.py .                                                                                                                                            [ 51%] py36/lib/python3.6/site-packages/glom/test/test_grouping.py ......                                                                                                                                   [ 54%] py36/lib/python3.6/site-packages/glom/test/test_let_vars.py .....                                                                                                                                    [ 56%] py36/lib/python3.6/site-packages/glom/test/test_match.py ........................                                                                                                                    [ 67%] py36/lib/python3.6/site-packages/glom/test/test_mutation.py ..................                                                                                                                       [ 75%] py36/lib/python3.6/site-packages/glom/test/test_path_and_t.py ..................                                                                                                                     [ 83%] py36/lib/python3.6/site-packages/glom/test/test_reduction.py .........                                                                                                                               [ 87%] py36/lib/python3.6/site-packages/glom/test/test_spec.py ..                                                                                                                                           [ 88%] py36/lib/python3.6/site-packages/glom/test/test_streaming.py .............                                                                                                                           [ 93%] py36/lib/python3.6/site-packages/glom/test/test_target_types.py ............                                                                                                                         [ 99%] py36/lib/python3.6/site-packages/glom/test/test_tutorial.py ..                                                                                                                                       [100%]

================================================================================== 224 passed, 2 skipped in 2.79 seconds ===================================================================================_________________________________________________________________________________________________ summary __________________________________________________________________________________________________
  py36: commands succeeded
  congratulations :)
```

**Comment by kurtbrose** [comment]
lol @ coverage error -- it's in chainmap_backport.py

**Comment by kurtbrose** [comment]
wellp... crap now that error is happening but python 3.5 not 3.6

**Comment by kurtbrose** [comment]
there's some rando exception happening apparently RARELY in the exception `__str__()` and I have no freaking clue where because it just says `<unprintable>`

**Comment by kurtbrose** [comment]
image
I put in a try / except that returned the traceback instead of the original exception...... and now the whole process crashes?

WTF is happening?

**Comment by kurtbrose** [comment]
1- re-ordered branches above error; this more accurately reflects the order things happened, and makes it easier to see where branches are coming from

2- added ASCII art `+` and `|` to associate a spec with its failed branches

3- made branches only use `|` instead of `  ` -- the `|` is strong enough that we can save horizontal space

4- added space in front of `Failed Branch` so it lines with `Spec` that is is associated with, also easy to read with some space between `|` and `F`

```python
Traceback (most recent call last):
  File "test_error.py", line ___, in _make_stack
    glom(target, spec)
  File "core.py", line ___, in glom
    raise err
glom.core.PathAccessError: error raised while processing, details below.
 Target-spec trace (most recent last):
 - Target: [None]
 - Spec: Match(Switch([(1, 1), ('a', 'a'), ([None], Switch([(1, 1), ('a', '...
 + Spec: Switch([(1, 1), ('a', 'a'), ([None], Switch([(1, 1), ('a', 'a'), (...
 | Failed Branch:
 |- Spec: 1
 |- glom.matching.MatchError: [None] does not match 1
 | Failed Branch:
 |- Spec: 'a'
 |- glom.matching.MatchError: [None] does not match 'a'
 | Failed Branch:
 |- Spec: [None]
 |+ Spec: Switch([(1, 1), ('a', 'a'), ([None], T.a)])
 || Failed Branch:
 ||- Spec: 1
 ||- glom.matching.MatchError: [None] does not match 1
 || Failed Branch:
 ||- Spec: 'a'
 ||- glom.matching.MatchError: [None] does not match 'a'
 || Failed Branch:
 ||- Spec: [None]
 ||- Spec: T.a
glom.core.PathAccessError: could not access 'a', part 0 of T.a, got error: AttributeError("'list' object has no attribute 'a'")
```

**Comment by kurtbrose** [comment]
experimenting with checks to mark branches

```python
 Target-spec trace (most recent last):
 - Target: [None]
 - Spec: Match(Switch([(1, 1), ('a', 'a'), ([None], Switch([(1, 1), ('a', '...
 + Spec: Switch([(1, 1), ('a', 'a'), ([None], Switch([(1, 1), ('a', 'a'), (...
 > Failed Branch:
 |- Spec: 1
 |- glom.matching.MatchError: [None] does not match 1
 > Failed Branch:
 |- Spec: 'a'
 |- glom.matching.MatchError: [None] does not match 'a'
 > Failed Branch:
 |- Spec: [None]
 |+ Spec: Switch([(1, 1), ('a', 'a'), ([None], T.a)])
 |> Failed Branch:
 ||- Spec: 1
 ||- glom.matching.MatchError: [None] does not match 1
 |> Failed Branch:
 ||- Spec: 'a'
 ||- glom.matching.MatchError: [None] does not match 'a'
 |> Failed Branch:
 ||- Spec: [None]
 ||- Spec: T.a
glom.core.PathAccessError: could not access 'a', part 0 of T.a, got error: AttributeError("'list' object has no attribute 'a'")
```

**Comment by mahmoud** [comment]
I've got a design over here with a hack implementation I don't dare commit: https://gist.github.com/mahmoud/ce593a4736b9d410669a3dbf5492950e

But basically it uses slashes to visually branch instead of the repetitive "Failed branch" text:

```

Traceback (most recent call last):
  File "tmp.py", line 9, in _make_stack
    glom(target, spec)
  File "/home/mahmoud/projects/glom/glom/core.py", line 2068, in glom
    raise err
glom.core.CoalesceError: error raised while processing, details below.
 Target-spec trace (most recent last):
 - Target: {'a': {'b': 'c'}, 'd': {'e': ['f'], 'g': 'h'}, 'i': [{'j... (len=5)
 + Spec: Coalesce(('xxx', 'z', 'n'), 'yyy')
 |\ Spec: ('xxx', 'z', 'n')
 || Spec: 'xxx'
 || Target: {'z': {'v': 0}}
 || Spec: 'z'
 || Target: {'v': 0}
 || Spec: 'n'
 |X glom.core.PathAccessError: could not access 'n', part 0 of Path('n'), got error: KeyError('n')
 |\ Spec: 'yyy'
 |X glom.core.PathAccessError: could not access 'yyy', part 0 of Path('yyy'), got error: KeyError('yyy')
glom.core.CoalesceError: no valid values found. Tried (('xxx', 'z', 'n'), 'yyy') and got (PathAccessError, PathAccessError) (at path ['xxx', 'z'])
```

**Comment by mahmoud** [comment]
- [ ] Possible flaky test (added -vv for more debug info)
- [ ] Add "reading branched exceptions" section to doc, maybe update CoalesceError
- [ ] See about ignoring chainmap if it isn't already

Could add more tests for different shapes.

**Comment by mahmoud** [comment]
docs changes are all in, but this test is definitely flaky. getting [unprintable exceptions](https://travis-ci.org/github/mahmoud/glom/jobs/712841629) on older py3s, which makes me think maybe the repr is dangerously close to the recursion limit or something? Even that doesn't strike me as nondeterministic...

Maybe a test with a really big/deep spec is in order, just to confirm some of the limits of the approach?

**Comment by mahmoud** [comment]
Gonna see about monkeypatching some debugging stuff into traceback for the next time the test flake repros.

**Comment by mahmoud** [comment]
OK, per our discussion, the CI visibility changes are in. If Travis acts up, we'll have some visibility. With that in place, I think this is pretty much ready to merge.

**Pull request merged**